src/states/node.rs

```diff
@@ -2096,7 +2123,25 @@ impl Node {
                    self,
                    pub_id);
             let _ = self.crust_service.disconnect(pub_id);
-            let _ = self.peer_mgr.remove_peer(pub_id);
+            if let Some((peer, _)) = self.peer_mgr.remove_peer(pub_id) {
+                match *peer.state() {
+                    PeerState::Bootstrapper(_) |
+                    PeerState::Client => {
+                        let _ = self.dropped_clients.insert(*pub_id, ());
+                    }
+                    PeerState::ConnectionInfoPreparing { .. } |
+                    PeerState::ConnectionInfoReady(_) |
+                    PeerState::CrustConnecting |
+                    PeerState::Connected(_) |
+                    PeerState::SearchingForTunnel |
+                    PeerState::JoiningNode |
+                    PeerState::Routing(_) |
+                    PeerState::Candidate(_) |
+                    PeerState::Proxy => (),
+                }
+            } else {
+                let _ = self.dropped_clients.insert(*pub_id, ());
```

Comment:
If a malicious client connects to us then sends a pile of traffic then drops the connection, this stops us handling any enqueued messages from it after we've called `disconnect_peer()` and removed it from `peer_mgr`.  (We've not yet handled the `LostPeer` event, but it's also in the queue after the malicious messages)